File: gps-track-visualizer.c
#include <stdbool.h>
#include <string.h>
#include <dirent.h>
#include <sys/stat.h>
#include <libxml/xmlreader.h>
#include <proj_api.h>
#include <gd.h>


enum { IMG_WIDTH  = 12800 };
enum { IMG_HEIGHT = 12800 };
enum { IMG_BORDER = 20 };
enum { MAX_POINTS = 1000000 };

struct point
{
	unsigned x, y;
	bool is_end;
};

static int count;
static int last_count;
static double total_length;

static double lon[MAX_POINTS];
static double lat[MAX_POINTS];

static struct point points[MAX_POINTS];


static void get_point_coordinates(xmlNodePtr trkpt)
{
	xmlChar *xml_lon = xmlGetProp(trkpt, (const xmlChar *) "lon");
	xmlChar *xml_lat = xmlGetProp(trkpt, (const xmlChar *) "lat");

	lon[count]   = atof((const char *) xml_lon);
	lat[count++] = atof((const char *) xml_lat);

	xmlFree(xml_lon);
	xmlFree(xml_lat);
}

static double distance(double lon1, double lat1, double lon2, double lat2)
{
	const int R = 6367500;

	lon1 *= M_PI / 180.0;
	lat1 *= M_PI / 180.0;
	lon2 *= M_PI / 180.0;
	lat2 *= M_PI / 180.0;

	double x = (lon2 - lon1) * cos((lat1 + lat2) / 2);
	double y = (lat2 - lat1);

	return sqrt(x * x + y * y) * R;
}

static double calc_length(int p1, int p2)
{
	double len = 0.0;

	double lon2 = lon[p1];
	double lat2 = lat[p1];
	for (int i = p1 + 1; i <= p2; i++)
	{
		double lon1 = lon2;
		double lat1 = lat2;
		lon2 = lon[i];
		lat2 = lat[i];

		if (!points[i - 1].is_end || !points[i].is_end)
			len += distance(lon1, lat1, lon2, lat2);
	}

	return len;
}

static int read_gpx(const char *gpx_filename)
{
	printf("Processing %s", gpx_filename);

	xmlDocPtr doc;
	if (!(doc = xmlParseFile(gpx_filename)))
	{
		fprintf(stderr, "XML document not parsed successfully\n");
		return 1;
	}

	xmlNodePtr root, trk, trkseg, trkpt;
	if (!(root = xmlDocGetRootElement(doc)))
	{
		fprintf(stderr, "Empty XML document\n");
		xmlFreeDoc(doc);
		return 1;
	}

	if (xmlStrcmp(root->name, (const xmlChar *) "gpx"))
	{
		fprintf(stderr, "XML document of the wrong type, root node != gpx\n");
		xmlFreeDoc(doc);
		return 1;
	}

	trk = root->xmlChildrenNode;
	while (trk != NULL)
	{
		if (!xmlStrcmp(trk->name, (const xmlChar *) "trk"))
		{
			trkseg = trk->xmlChildrenNode;
			while (trkseg != NULL)
			{
				points[count].is_end = true;
				if (!xmlStrcmp(trkseg->name, (const xmlChar *) "trkseg"))
				{
					trkpt = trkseg->xmlChildrenNode;
					while (trkpt != NULL)
					{
						if (!xmlStrcmp(trkpt->name, (const xmlChar *) "trkpt"))
							get_point_coordinates(trkpt);

						if (count >= MAX_POINTS)
							return 1;

						trkpt = trkpt->next;
					}
				}
				points[count - 1].is_end = true;
				trkseg = trkseg->next;
			}
		}
		trk = trk->next;
	}

	xmlFreeDoc(doc);

	double length = calc_length(last_count, count - 1);
	last_count = count;
	total_length += length;
	printf(" [%.2f km]\n", length / 1000);

	return 0;
}

static int lonlat_to_xy()
{
	projPJ pj_src, pj_dst;

	if (!(pj_src = pj_init_plus("+proj=latlong +ellps=WGS84 +datum=WGS84")))
		return 1;
	if (!(pj_dst = pj_init_plus("+proj=utm +ellps=WGS84 +lon_0=39"))) // Moscow
		return 1;

	for (int i = 0; i < count; i++)
	{
		lon[i] *= M_PI / 180.0;
		lat[i] *= M_PI / 180.0;
	}

	pj_transform(pj_src, pj_dst, count, 1, lon, lat, NULL);

	pj_free(pj_src);
	pj_free(pj_dst);

	double min_lon = lon[0];
	double max_lon = lon[0];
	double min_lat = lat[0];
	double max_lat = lat[0];

	for (int i = 1; i < count; i++)
	{
		min_lon = (min_lon < lon[i]) ? min_lon : lon[i];
		max_lon = (max_lon > lon[i]) ? max_lon : lon[i];
		min_lat = (min_lat < lat[i]) ? min_lat : lat[i];
		max_lat = (max_lat > lat[i]) ? max_lat : lat[i];
	}

	double k_lon = (IMG_WIDTH  - 2 * IMG_BORDER) / (max_lon - min_lon);
	double k_lat = (IMG_HEIGHT - 2 * IMG_BORDER) / (max_lat - min_lat);

	double k = (k_lon < k_lat) ? k_lon : k_lat;

	for (int i = 0; i < count; i++)
	{
		points[i].x = IMG_BORDER + k * (lon[i] - min_lon) + 0.5;
		points[i].y = IMG_BORDER + k * (lat[i] - min_lat) + 0.5;
	}

	return 0;
}

static int draw_img(const char *img_file_name)
{
	gdImagePtr img = gdImageCreate(IMG_WIDTH, IMG_HEIGHT);

	const int dot_r = 3;
	int black = gdImageColorAllocate(img, 0,   0, 0); // Set background color
	int green = gdImageColorAllocate(img, 0, 255, 0);
	int red   = gdImageColorAllocate(img, 255, 0, 0);

	struct point p2 = points[0];
	for (int i = 1; i < count; i++)
	{
		struct point p1 = p2;
		p2 = points[i];

		if (!p1.is_end || !p2.is_end)
			gdImageLine(img, p1.x, IMG_HEIGHT - p1.y, p2.x, IMG_HEIGHT - p2.y, green);
	}

	for (int i = 0; i < count; i++)
	{
		struct point p = points[i];
		if (p.is_end)
			gdImageFilledEllipse(img, p.x, IMG_HEIGHT - p.y, dot_r, dot_r, red);
	}

	FILE *pngout = fopen(img_file_name, "wb");
	if (!pngout)
	{
		perror(img_file_name);
		return 1;
	}

	gdImagePng(img, pngout);
	fclose(pngout);
	gdImageDestroy(img);

	return 0;
}

static int read_dir(const char *dir_name)
{
	DIR *dir = opendir(dir_name);
	if (!dir)
		return 1;

	struct dirent *ent;
	while ((ent = readdir(dir)))
		if (ent->d_name[0] != '.')
		{
			char track[256];
			strncpy(track, dir_name, sizeof(track) - 1);
			strncat(track, ent->d_name, sizeof(track) - strlen(track) - 1);

			if (read_gpx(track) != 0)
			{
				closedir(dir);
				return 2;
			}
		}

	closedir(dir);
	return 0;
}

int main(int argc, char *argv[])
{
	if (argc != 3)
	{
		fprintf(stderr, "Usage: %s track_list out_img\n", argv[0]);
		return 1;
	}

	const char *track_list = argv[1];
	const char *out_img    = argv[2];

	FILE *file = fopen(track_list, "r");
	if (!file)
	{
		perror(track_list);
		return 2;
	}

	char track[256];
	while (fgets(track, sizeof track, file) != NULL)
	{
		char *newline = strchr(track, '\n');
		if (newline)
			*newline = '\0';

		struct stat st;
		if (stat(track, &st) == 0 && S_ISDIR(st.st_mode))
		{
			if (read_dir(track) != 0)
			{
				fclose(file);
				return 3;
			}
		}
		else
		{
			if (read_gpx(track) != 0)
			{
				fclose(file);
				return 4;
			}
		}
	}
	fclose(file);

	printf("Total: %.2f km\n", total_length / 1000);

	if (lonlat_to_xy())
		return 5;

	if (draw_img(out_img))
		return 6;

	return 0;
}

File: gps-track-visualizer.py
#!/usr/bin/env python3

import os
import math
import argparse
import xml.etree.ElementTree as ET
from pyproj import Transformer, CRS
from pyproj.aoi import AreaOfInterest
from pyproj.database import query_utm_crs_info
from PIL import Image, ImageDraw

points = [ ]
total_length = 0


def distance(p1, p2):
    lon1 = math.radians(float(p1['lon']))
    lat1 = math.radians(float(p1['lat']))
    lon2 = math.radians(float(p2['lon']))
    lat2 = math.radians(float(p2['lat']))

    a = (lon2 - lon1) * math.cos((lat1 + lat2) / 2)
    b = (lat2 - lat1)
    r = 6363720 # Earth radius at ground level for latitude 55.751244

    return math.sqrt(a * a + b * b) * r


def read_gpx(gpx_filename):
    print("Processing %s" % gpx_filename, end='')

    try:
        tree = ET.ElementTree(file=gpx_filename)
    except FileNotFoundError as err:
        exit(err)
    except ET.ParseError as err:
        exit(err)

    ns = { 'gpx': 'http://www.topografix.com/GPX/1/1' }
    root = tree.getroot()
    if root.tag != '{%s}gpx' % ns['gpx']:
        exit("XML document of the wrong type, root node != gpx")

    trk_length = 0
    for trk in root.findall('gpx:trk', ns):
        for trkseg in trk.findall('gpx:trkseg', ns):
            trkseg_length = 0
            is_start = True                 # First point of the track segment

            for trkpt in trkseg.findall('gpx:trkpt', ns):
                p = trkpt.attrib
                p['terminal'] = is_start
                if not is_start:
                    trkseg_length = trkseg_length + distance(points[-1], p)
                points.append(p)
                is_start = False

            points[-1]['terminal'] = True   # Last point of the track segment
            trk_length = trk_length + trkseg_length

    print(" [%.2f km]" % (trk_length / 1000))
    global total_length
    total_length = total_length + trk_length


def read_dir(dir_name):
    for f in sorted(os.listdir(dir_name)):
        if not f.startswith('.'):
            f = os.path.join(dir_name, f)
            if os.path.isfile(f):
                read_gpx(f)


def lonlat_to_xy(img_width, img_height, img_border):
    print("Converting coordinates...")

    utm_crs_list = query_utm_crs_info(
        datum_name = "WGS 84",
        area_of_interest = AreaOfInterest(
            west_lon_degree  = min(float(p['lon']) for p in points),
            south_lat_degree = min(float(p['lat']) for p in points),
            east_lon_degree  = max(float(p['lon']) for p in points),
            north_lat_degree = max(float(p['lat']) for p in points)))

    transformer = Transformer.from_crs(
        {"proj":'latlong', "ellps":'WGS84', "datum":'WGS84'},
        CRS.from_epsg(utm_crs_list[0].code),
        always_xy=True)

    for p in points:
        p['x'], p['y'] = transformer.transform(p['lon'], p['lat'])

    min_x = min(p['x'] for p in points)
    max_x = max(p['x'] for p in points)
    min_y = min(p['y'] for p in points)
    max_y = max(p['y'] for p in points)
    k_x = (img_width  - 2 * img_border) / (max_x - min_x)
    k_y = (img_height - 2 * img_border) / (max_y - min_y)
    k = min(k_x, k_y)

    for p in points:
        p['x'] = round(img_border + k * (p['x'] - min_x))
        p['y'] = img_height - round(img_border + k * (p['y'] - min_y))


def draw_img(img_file_name, img_width, img_height):
    print("Creating image...")

    palette = [ 0,0,0, 0,255,0, 255,0,0 ]
    img = Image.new(mode='P', size=(img_width, img_height))
    img.putpalette(palette)
    draw = ImageDraw.Draw(img)

    p2 = points[0]
    for p in points[1:]:
        p1 = p2
        p2 = p
        if not (p1['terminal'] and p2['terminal']):
            draw.line((p1['x'], p1['y'], p2['x'], p2['y']), fill=1)

    for p in points:
        if p['terminal']:
            draw.line((p['x'] - 1, p['y'], p['x'] + 1, p['y']), fill=2)
            draw.line((p['x'], p['y'] - 1, p['x'], p['y'] + 1), fill=2)

    del draw
    img.save(img_file_name)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('track_list', help="list of GPX tracks or directories with tracks")
    parser.add_argument('out_image',  help="output PNG image")
    parser.add_argument('--width',  type=int, default=12800, help="width of the output image")
    parser.add_argument('--height', type=int, default=12800, help="height of the output image")
    parser.add_argument('--border', type=int, default=20,    help="border on the output image")
    args = parser.parse_args()

    try:
        f = open(args.track_list)
    except FileNotFoundError as err:
        exit(err)

    for track in f:
        track = track.rstrip()
        if os.path.isdir(track):
            read_dir(track)
        else:
            read_gpx(track)
    f.close()

    print("Total: %.2f km\n" % (total_length / 1000))

    lonlat_to_xy(args.width, args.height, args.border)

    draw_img(args.out_image, args.width, args.height)


if __name__ == '__main__':
    main()
